### PR TITLE
KAFKA-3723 Cannot change size of schema cache for JSON converter

### PR DESCRIPTION
**Pull request by christian-posta**

### CODE REVIEW

**Comment by christian-posta** [comment]
@ewencp would be great if you could take a look at this!

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java [hunk 1]
```diff
@@ -304,7 +304,7 @@ public void configure(Map<String, ?> configs, boolean isKey) {
 
         Object cacheSizeVal = configs.get(SCHEMAS_CACHE_SIZE_CONFIG);
         if (cacheSizeVal != null)
-            cacheSize = (int) cacheSizeVal;
+            cacheSize = Integer.parseInt((String) cacheSizeVal);
```
Comment (ewencp) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:307]: Hmm, so actually I think the problem here is that we need to be able to parse generally because these can technically be passed in as `Map<String, ?>`, so the type is unknown.

Another option is to replace this parsing with `ConfigDef`, which will handle both the `String` and `Integer` cases since its designed to handle both properties files and `Maps` passed in directly from code. To be honest, I think the only reason we didn't do this before is because this code came with the initial patch that was prototyped outside of Kafka and couldn't use internal APIs.

File: connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java [hunk 2]
```diff
@@ -607,6 +611,19 @@ public void testCacheSchemaToJsonConversion() {
         assertEquals(2, cache.size());
     }
 
+    @Test
+    public void testJsonSchemaCacheSizeFromConfigFile() throws URISyntaxException, IOException {
+        URL url = getClass().getResource("/connect-test.properties");
```
Comment (ewencp) [inline connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:616]: This works fine, though if you wanted to simplify, the key thing here is just that you are using a `Map<String, String>`, so you could get rid of the resource file and just create it on the fly:

```
Collections.singletonMap("schemas.cache.size", "1");
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java [hunk 3]
```diff
@@ -57,6 +57,10 @@
     public static final String VALUE_CONVERTER_CLASS_DOC =
             "Converter class for value Connect data that implements the <code>Converter</code> interface.";
 
+    public static final String VALUE_CONVERTER_SCHEMA_CACHE_SIZE_CONFIG = "value.converter.schemas.cache.size";
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: The intent with all the pluggable components in connect is to follow the same pattern as the clients wrt config. So we actually don't want to expose these configs via WorkerConfig. If we did, we'd basically be committing all converters to honor these configs. This cache one might make sense generally, but enable/disable schemas below definitely doesn't.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java [hunk 4]
```diff
@@ -81,9 +81,7 @@
         HERDER_CONFIG.put(WorkerConfig.VALUE_CONVERTER_CLASS_CONFIG, "org.apache.kafka.connect.json.JsonConverter");
         HERDER_CONFIG.put(WorkerConfig.INTERNAL_KEY_CONVERTER_CLASS_CONFIG, "org.apache.kafka.connect.json.JsonConverter");
         HERDER_CONFIG.put(WorkerConfig.INTERNAL_VALUE_CONVERTER_CLASS_CONFIG, "org.apache.kafka.connect.json.JsonConverter");
-        HERDER_CONFIG.put(DistributedConfig.CONFIG_TOPIC_CONFIG, "connect-configs");
```
Comment (ewencp) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:84]: Why are these removed? If this still passes, I'm surprised -- these are required configs in DistributedConfig.

File: connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java (hunk 2 above)
Comment (christian-posta) [inline connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:616]: +1.. yah i get that.. but wanted to simulate exactly what happens when we configure this from a file. I'll change it if you prefer.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java (hunk 3 above)
Comment (christian-posta) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Okay I can change that. I was hoping to avoid magic properties that don't appear anywhere in the docs and isn't clear what the props should be.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java (hunk 4 above)
Comment (christian-posta) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:84]: they are dups. tests run fine w/o them

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java (hunk 1 above)
Comment (christian-posta) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:307]: Actually, they're all casted explicitly to `Map<String,String>`

https://github.com/christian-posta/kafka/blob/c9485b78a6e43747daf1314ae9532839fb7bc810/connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java#L61-61

https://github.com/christian-posta/kafka/blob/18f643b82b05904516295e2734af0d8a9fb053ea/connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectStandalone.java#L66-66

And both StandaloneConfig and DistributedConfig accept only `Map<String,String>`

https://github.com/christian-posta/kafka/blob/079c88178dff4b3a4c9de55629e7d15b60e5f562/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java#L197-197

https://github.com/christian-posta/kafka/blob/079c88178dff4b3a4c9de55629e7d15b60e5f562/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneConfig.java#L42-42

I tried change them all to `Map<String,Object>` but that caused other problems.

Thoughts?

**Comment by christian-posta** [comment]
I'd love additional guidance on this so we can commit it?

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java (hunk 1 above)
Comment (ewencp) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:307]: Uggh, yeah, I forgot about this. We kind of inherit some annoying types from Kafka's config setup but I tried to ensure we're using String types where possible. It gets a bit hard to figure out what is valid where -- the JsonConverter actually gets passed a `Map<String, Object>` as that's what is returned by `AbstractConfig.originalsWithPrefix`, but in practice they are all `String` so type erasure allows this to work...

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java (hunk 3 above)
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Yeah, I agree its not ideal. Unfortunately the way Kafka normally does configs does not leave room for a better solution where the different configuration sets can be clearly separated so its clear where you need to look for config options. And agree its not great having to look in multiple locations, but that is the reality of a pluggable system.

**Comment by ewencp** [comment]
@christian-posta If we remove the `value.converter.` prefixed configs, this LGTM

**Comment by christian-posta** [comment]
@ewencp updated the PR. thank you for the review!

**Comment by ewencp** [comment]
LGTM, thanks!
